### PR TITLE
Optional fip

### PR DESCRIPTION
**Pull request by mkjpryor-stfc**

### CODE REVIEW

File: group_vars/openstack/k8s [hunk 1]
```diff
@@ -29,9 +18,12 @@ cluster_workers:
   flavor: "{{ cluster_worker_flavor | default('j3.medium') }}"
   root_volume_size: 0
 
-cluster_groups_without_gw:
-  - "{{ (cluster_gw_group == 'kube-masters') | ternary(cluster_masters_with_fip, cluster_masters) }}"
-  - "{{ cluster_workers }}"
+k8s_groups_fixed_ip:
+  - "{{ k8s_group_masters | combine(cluster_gw_fip_mixin) }}"
```
Comment (brtkwr) [inline group_vars/openstack/k8s:22]: I like this!

File: tasks/infra/provision.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+
+# If no fixed IP is given, see if there is an existing one from an ephemeral deployment gateway
```
Comment (brtkwr) [inline tasks/infra/provision.yml:3]: Hmm lets home this pattern works for all infra type :)

**Review by brtkwr — APPROVED** [review]
Looks good, the only thing I'm concerned about is whether tasks/infra/provision.yml works with all infra types... but in theory it should. We'll have to apply it and see what happens.

File: nfs-infra.yml [hunk 3]
```diff
@@ -1,47 +1,37 @@
...
-  # Hack pending outcome of:
-  # - https://github.com/ansible/ansible/pull/53757
-  # - https://review.openstack.org/643195
   tasks:
-    - name: Fixup stack tags
-      command: "openstack stack update --existing --tags {{ cluster_type.storage }} {{ cluster_name }}"
+    - import_tasks: tasks/infra/cleanup-gateway.yml
+      vars:
+        cluster_groups: "{{ nfs_groups }}"
+        cluster_tag: "{{ cluster_type.storage }}"
```
Comment (brtkwr) [inline nfs-infra.yml:37]: Looks good!

**Comment by mkjpryor-stfc** [comment]
@brtknr I had it like that originally and it complained :P

**Comment by mkjpryor-stfc** [comment]
@brtknr It complained because the `'some string'` is valid YAML syntax, so it didn't know what to do with the `in groups`

**Comment by brtkwr** [comment]
Thats odd!

**Comment by brtkwr** [comment]
Are you sure then that it is not resolving as true because it is non-empty string rather than meeting the condition?

**Comment by mkjpryor-stfc** [comment]
No - it is a parse error. It doesn't even get as far as executing any tasks.

**Comment by mkjpryor-stfc** [comment]
```
ERROR! Syntax Error while loading YAML.
  did not find expected key

The error appears to have been in '/Users/luj83518/dev/cluster-as-a-service/jasmin-appliances/nfs-infra.yml': line 19, column 22, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

      loop: "{{ groups['server'] }}"
      when: 'server' in groups
                     ^ here
This one looks easy to fix.  It seems that there is a value started
with a quote, and the YAML parser is expecting to see the line ended
with the same kind of quote.  For instance:

    when: "ok" in result.stdout

Could be written as:

   when: '"ok" in result.stdout'

Or equivalently:

   when: "'ok' in result.stdout"
```

**Pull request merged**